tests/test_experimental.py

```diff
@@ -1,14 +1,392 @@
+
+    rng = np.random.default_rng(seed)
+    rows = rng.integers(10, 100)
+    columns = rng.integers(2, 9)
+
+    # Create data matrix
+    X = rng.integers(0, 10, size=(rows, columns))
+
+    groups = list(groupby_indices(X))
+    indices = [set(idx) for (_, idx) in groups]
+
+    # Verify that every row is included
+    union_idx = functools.reduce(set.union, indices)
+    assert union_idx == set(range(rows))
+
+    # Verify that no duplicate rows occur
+    intersection_idx = functools.reduce(set.intersection, indices)
+    assert intersection_idx == set()
+
+    # Verify each entry in the groups
+    for (unique_row, indices_of_row) in groups:
+
+        # Repeat this unique row the number of times it occurs in X
+        repeated = np.repeat(
+            unique_row[np.newaxis, :], repeats=len(indices_of_row), axis=0
+        )
+        assert np.allclose(X[indices_of_row, :], repeated)
+
+
+@pytest.fixture()
+def linear_problem(request):
+
+    # Seed the problem using indirect parametrization:
+    # https://docs.pytest.org/en/latest/example/parametrize.html#indirect-parametrization
+    rng = np.random.default_rng(request.param)
+
+    # Create a problem with g(x) = A @ x
+    num_parameters = 50
+    num_observations = 10
+    num_ensemble = 100
+
+    A = np.exp(rng.standard_normal(size=(num_observations, num_parameters)))
+
+    def g(X):
+        """Forward model."""
+        return A @ X
+
+    # Create observations
+    x_true = np.linspace(-1, 1, num=num_parameters)
+    observations = g(x_true) + rng.standard_normal(size=num_observations) / 10
+
+    # Initial ensemble and covariance
+    X = rng.normal(size=(num_parameters, num_ensemble))
+    covariance = rng.triangular(0.1, 1, 1, size=num_observations)
+    yield X, g, observations, covariance, rng
+
+
+class TestAdaptiveESMDA:
+    @pytest.mark.parametrize(
+        "linear_problem",
+        range(25),
+        indirect=True,
+        ids=[f"seed-{i+1}" for i in range(25)],
+    )
+    def test_that_adaptive_localization_with_cutoff_1_equals_ensemble_prior(
+        self, linear_problem
+    ):
+
+        # Create a problem with g(x) = A @ x
+        X, g, observations, covariance, _ = linear_problem
+
+        # Create adaptive smoother
+        smoother = AdaptiveESMDA(
+            covariance=covariance, observations=observations, seed=1
+        )
+
+        X_i = np.copy(X)
+        alpha = normalize_alpha(np.ones(5))
+        for alpha_i in alpha:
+
+            # Run forward model
+            Y_i = g(X_i)
+
+            # Create noise D - common to this ESMDA update
+            D_i = smoother.perturb_observations(
+                ensemble_size=Y_i.shape[1], alpha=alpha_i
+            )
+
+            # Update the relevant parameters and write to X (storage)
+            X_i = smoother.assimilate(
+                X=X_i,
+                Y=Y_i,
+                D=D_i,
+                alpha=alpha_i,
+                correlation_threshold=lambda ensemble_size: 1,
```

Comment:
Would it be good if we could set `correlation_threshold` as a `float` as well as a `Callable`?
Seems kind of natural to call `assimilate` with `correlation_threshold=1.0`.

Reply:
I've added this, but I am on the fence. It's easy enough to add, so I did it.